src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs

```diff
@@ -1416,28 +1433,139 @@ private void CompileSwitchExpression(Expression expr)
+                        CompileStringSwitchExpression(node);
+                        return;
+                    }
+                }
+            }
+
+            LocalDefinition temp = _locals.DefineLocal(Expression.Parameter(node.SwitchValue.Type), _instructions.Count);
+            Compile(node.SwitchValue);
+            _instructions.EmitStoreLocal(temp.Index);
+                        
+            var doneLabel = Expression.Label(node.Type, "done");
+
+            foreach(var @case in node.Cases)
+            {
+                foreach(var val in @case.TestValues)
+                {
+                    //  temp == val ? 
+                    //          goto(Body) doneLabel: 
+                    //          {};
+                    CompileConditionalExpression(
+                        Expression.Condition(
+                            Expression.Equal(temp.Parameter, val, false, node.Comparison),
+                            Expression.Goto(doneLabel, @case.Body),
+                            AstUtils.Empty()
+                        ),
+                        asVoid: true);
+                }
+            }
+
+            // doneLabel(DefaultBody):
+            CompileLabelExpression(Expression.Label(doneLabel, node.DefaultBody));
+
+            _locals.UndefineLocal(temp, _instructions.Count);
+        }
+
+        private void CompileIntSwitchExpression<T>(SwitchExpression node)
+        {
+            LabelInfo end = DefineLabel(null);
+            bool hasValue = node.Type != typeof(void);
+
+            Compile(node.SwitchValue);
+            var caseDict = new Dictionary<T, int>();
+            int switchIndex = _instructions.Count;
+            _instructions.EmitIntSwitch(caseDict);
+
+            if (node.DefaultBody != null)
             {
-                throw new PlatformNotSupportedException(SR.SwitchOverNonInt32ValuesNotSupported);
+                Compile(node.DefaultBody);
             }
-            if (node.Comparison != null)
+            else
             {
-                throw new PlatformNotSupportedException(SR.SwitchWithCustomEqualityComparisonMethodNotSupported);
+                Debug.Assert(!hasValue);
             }
+            _instructions.EmitBranch(end.GetLabel(this), false, hasValue);
 
-            // Test values must be constant
-            if (!node.Cases.All(c => c.TestValues.All(t => t is ConstantExpression)))
+            for (int i = 0; i < node.Cases.Count; i++)
             {
-                throw new PlatformNotSupportedException(SR.NonConstantCaseValuesInSwitchNotSupported);
+                var switchCase = node.Cases[i];
+
+                int caseOffset = _instructions.Count - switchIndex;
+                foreach (ConstantExpression testValue in switchCase.TestValues)
+                {
+                    var key = (T)testValue.Value;
+                    if (!caseDict.ContainsKey(key))
+                    {
+                        caseDict.Add(key, caseOffset);
```

Comment:
I think this was proposed many times, in particular since concurrent dictionary has this. Not sure why that was not done. Perhaps because IDictionary does not have this and it would be breaking to add?